test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java

```diff
@@ -0,0 +1,268 @@
+                //WEB-INF/lib -> lib/
+                if (Files.exists(tmpLocation.resolve("WEB-INF/lib"))) {
+                    Files.move(tmpLocation.resolve("WEB-INF/lib"), tmpLocation.resolve("lib"));
+                }
+                //WEB-INF/classes -> archive/
+                if (Files.exists(tmpLocation.resolve("WEB-INF/classes"))) {
+                    Files.move(tmpLocation.resolve("WEB-INF/classes"), appLocation);
+                } else {
+                    Files.createDirectory(appLocation);
+                }
+                //META-INF -> archive/META-INF/
+                if (Files.exists(tmpLocation.resolve("META-INF"))) {
+                    Files.move(tmpLocation.resolve("META-INF"), appLocation.resolve("META-INF"));
+                }
+                //WEB-INF -> archive/WEB-INF
+                if (Files.exists(tmpLocation.resolve("WEB-INF"))) {
+                    Files.move(tmpLocation.resolve("WEB-INF"), appLocation.resolve("WEB-INF"));
+                }
+                // Collect all libraries
+                if (Files.exists(tmpLocation.resolve("lib"))) {
+                    Files.walk(tmpLocation.resolve("lib"), 1).forEach(libraries::add);
+                }
+            } else {
+                appLocation = tmpLocation;
+            }
+
+            List<Consumer<BuildChainBuilder>> customizers = new ArrayList<>();
+            try {
+                // Test class is a bean
+                Class<? extends BuildItem> buildItem = Class
+                        .forName("io.quarkus.arc.deployment.AdditionalBeanBuildItem").asSubclass(BuildItem.class);
+                customizers.add(new Consumer<BuildChainBuilder>() {
+                    @Override
+                    public void accept(BuildChainBuilder buildChainBuilder) {
+                        buildChainBuilder.addBuildStep(new BuildStep() {
+                            @Override
+                            public void execute(BuildContext context) {
+                                try {
+                                    Method factoryMethod = buildItem.getMethod("unremovableOf", Class.class);
+                                    context.produce((BuildItem) factoryMethod.invoke(null, testJavaClass));
+                                } catch (Exception e) {
+                                    throw new RuntimeException(e);
+                                }
+                            }
+                        }).produces(buildItem)
+                                .build();
+                    }
+                });
+            } catch (ClassNotFoundException e) {
+                throw new IllegalStateException(e);
+            }
+
+            Path testClassesLocation;
+            try {
+                testClassesLocation = PathTestHelper.getTestClassesLocation(testJavaClass);
+            } catch (Exception e) {
+                // TCK tests are usually located in a dependency jar
```

Comment:
Is it worth logging a warning here? Are we sure every exception can be handled by falling back to `target/test-classes`?